HW2/P5/P5.txt

```diff
@@ -0,0 +1,137 @@
+
+4 Threads
+1309.86412287 simulation frames per second
+1315.05369969 simulation frames per second
+1224.0477187 simulation frames per second
+1073.40185837 simulation frames per second
+1047.48362422 simulation frames per second
+885.646555365 simulation frames per second
+1079.32939234 simulation frames per second
+1051.8630721 simulation frames per second
+563.354729998 simulation frames per second
+1150.65211978 simulation frames per second
+1076.89689258 simulation frames per second
+1034.55774259 simulation frames per second
+
+Analysis
+~~~~~~~~
+Introducing the grid yields some incredible results. On one thread, the speedup is approximately 100x to a range of 550-730 and on 4 threads the speedup is a bit less, around 80x from the 4-threaded gridless version. This speedup is expected since the algorithm is linear instead of polynomial. At the same time, on the small test simulation, it seems to be that the balls overlap more frequently and to a greater degree than they did without the grid. Though I cannot confirm this in any systematic manner, it may be the case that we gain a performance boost by trading off collision detection accuracy. Ultimately though, the combination of algorithm optimization and multithreading boosts performance very noticeably (though the algorithm improvement clearly holds more weight).
+
+Problem 3. Sorting
+----------------
+Results
+~~~~~~~
+1 Thread
+701.579275305 simulation frames per second
+903.706084321 simulation frames per second
+872.601386963 simulation frames per second
+872.262644188 simulation frames per second
+905.827770463 simulation frames per second
+871.061022464 simulation frames per second
+869.702240844 simulation frames per second
+900.501585982 simulation frames per second
+873.503967345 simulation frames per second
+779.92822354 simulation frames per second
+
+4 Threads
+1644.55540268 simulation frames per second
+2649.96809757 simulation frames per second
+2404.22495382 simulation frames per second
+1709.7589452 simulation frames per second
+2251.73668633 simulation frames per second
+2512.88486941 simulation frames per second
+2345.35929614 simulation frames per second
+1996.52136351 simulation frames per second
+1998.60866102 simulation frames per second
+
+Analysis
+~~~~~~~~
+For this problem, I used Morton coding, as I was familiar with it from CS207 and know that it is generally a good way to flatten coordinates in 2 dimensions into a single curve. I used bitshift operations found on a Stanford course webpage to compute the interleavings of the coordinates and used these interleavings as the keys to sort the coordinates appropriately.
+The efficacy of the sorting surprised me. I understand that since my collision detection checks the 5x5 bounding box of a single ball's grid space, we only ever check balls that are near our target ball. This means that if they are sorted closer together in memory, we can expect to experience marginal improvements in the collision detection.
```

Comment:
Since you have done this before, do you have any insight on why Morton may be better than the Hildebert? I was under the impression that because it has less operations it may yield less overhead, but I am not sure.